ext/ruby_newt/ruby_newt.c

```diff
@@ -536,19 +537,106 @@ static VALUE rb_ext_Widget_takesFocus(VALUE self, VALUE index)
   return Qnil;
 }
 
-static VALUE rb_ext_Widget_equal(VALUE self, VALUE widget)
+static VALUE rb_ext_Widget_equal(VALUE self, VALUE obj)
 {
-  newtComponent co, co2;
+  newtComponent co;
+  void *data;
 
-  if (NIL_P(widget)) return Qfalse;
-  if (self == widget) return Qtrue;
+  if (NIL_P(obj)) return Qfalse;
+  if (self == obj) return Qtrue;
+
+  if (rb_obj_is_kind_of(obj, cWidget) || rb_obj_is_kind_of(obj, cExitStruct)) {
+    Data_Get_Struct(self, struct newtComponent_struct, co);
+    Data_Get_Struct(obj, void, data);
+    if (co == data) return Qtrue;
+    if (rb_obj_is_kind_of(obj, cExitStruct)
+      && co == ((struct newtExitStruct *) data)->u.co) return Qtrue;
```

Comment:
I don't think it matters. I'm assuming that if the data pointers are the same, then the other object is also a cWidget. Unless I'm missing something?